dataloader.py

```diff
@@ -0,0 +1,145 @@
+        text_tokens = self._extract_text_tokens(text)
+
+        # Extract audio waveform
+        wav = self._extract_audio_waveform(audio_path)
+
+        # Extract audio token extraction
+        # audio_tokens = self._extract_audio_tokens(audio_path)
+
+        # Extract speaker embedding
+        speaker_embedding = self._extract_speaker_embeddings(audio_path)
+
+
+        # Some of these fields may be redundant, useful for testing right now.
+        return text_tokens, wav, speaker_embedding, text
+
+    def _extract_text_tokens(self, text: str):
+        # For text tokens, one can use the tokenizer per:
+        # https://github.com/metavoiceio/metavoice-src/blob/main/fam/llm/inference.py#L177
+        encoded = encode_tokens(self.tokenizer, text, device=self.device)
+
+        return encoded
+
+    def _extract_audio_waveform(self, audio_path: str):
+        # read audio
+        wav, sr = audio_read(audio_path)
+
+        # Resample to MBD's expected sample rate
+        if sr != MBD_SAMPLE_RATE:
+            wav = julius.resample_frac(wav, sr, MBD_SAMPLE_RATE)
+
+        # Convert to mono and fix dimensionality
+        if wav.ndim == 2:
+            wav = wav.mean(axis=0, keepdims=True)
+        
+        wav = wav.unsqueeze(0)  # Add batch dimension
+        wav = wav.unsqueeze(0)  # Add channel dimension
+
+        return wav
+
+    # def _extract_audio_tokens(self, audio_path: str):
+    #     # read audio
+    #     wav, sr = audio_read(audio_path)
+
+    #     # Resample to MBD's expected sample rate
+    #     if sr != MBD_SAMPLE_RATE:
+    #         wav = julius.resample_frac(wav, sr, MBD_SAMPLE_RATE)
+
+    #     # Convert to mono and fix dimensionality
+    #     if wav.ndim == 2:
+    #         wav = wav.mean(axis=0, keepdims=True)
+    #     wav = wav.unsqueeze(0)  # Add batch dimension
+
+    #     # Extract tokens
+    #     wav = wav.to(self.device)
+    #     tokens = self.encodec_model.encode(wav) # list[EncodedFrame = tp.Tuple[torch.Tensor, tp.Optional[torch.Tensor]]]
+
+    #     tokens = tokens[0][0][0] # shape (8, T)
+    #     print("tokens2 shape ", tokens.shape)
+
+    #     # Only return tokens in first 2 hierarchies for training stage 1
+    #     # Not sure if this is the correct approach.
+    #     tokens = tokens[:2]
+    #     print("tokens3 shape ", tokens.shape) # shape (2, T)
+        
+    #     return tokens
+
+    def _extract_speaker_embeddings(self, audio_path: str):
+        # For speaker embedding, you can also follow the code at:
+        # https://github.com/metavoiceio/metavoice-src/blob/main/fam/llm/inference.py#L435
+        return get_cached_embedding(audio_path, self.spkemb_model)
+
+def custom_collate_fn(batch):
+    text_tokens, wavs, speaker_embeddings, texts = zip(*batch)
+    
+    # Padding for text tokens
+    text_tokens = pad_sequence(text_tokens, batch_first=True, padding_value=0)
+    
+    # Audio waveform - padding to longest waveform on 4th dimension
```

Comment:
Why do we have 4 dimension? Should be either (batch_size, time) or (batch_size, channels=1, time)?

Reply:
Since we only need the Encodec tokens for training stage 1, I have removed the audio waveforms from the dataloader.